src/credentials/GroupDataProvider.h

```diff
@@ -0,0 +1,178 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/util/basic-types.h>
+#include <lib/core/CHIPError.h>
+#include <stdint.h>
+#include <sys/types.h>
+
+namespace chip {
+namespace Credentials {
+
+class GroupDataProvider
+{
+public:
+    static constexpr size_t GroupNameLengthMax = 16;
+    static constexpr size_t GroupEndpointsMax  = 16;
+
+    struct EpochKey
+    {
+        static constexpr size_t kLengthBytes = 16;
+        uint64_t start_time;
+        uint8_t key[kLengthBytes];
+    };
+
+    struct GroupMapping
+    {
+        EndpointId endpoint;
+        GroupId group;
+        bool operator==(const GroupMapping & other) { return this->endpoint == other.endpoint && this->group == other.group; }
+        GroupMapping & operator=(const GroupMapping & other)
+        {
+            if (this == &other)
+                return *this; // Guard self assignment
+            this->endpoint = other.endpoint;
+            this->group    = other.group;
+            return *this;
+        }
+    };
+
+    struct GroupState
+    {
+        uint16_t state_index;
+        chip::GroupId group;
+        uint16_t key_set_index;
+        bool operator==(const GroupState & other)
+        {
+            return this->state_index == other.state_index && this->group == other.group &&
+                this->key_set_index == other.key_set_index;
+        }
+    };
+
+    struct KeySet
+    {
+        enum class SecurityPolicy : uint8_t
+        {
+            kStandard   = 0,
+            kLowLatency = 1
+        };
+
+        uint16_t key_set_index;
+        SecurityPolicy policy;
+        EpochKey epoch_keys[3];
+    };
+
+    class GroupMappingIterator
+    {
+    public:
+        virtual ~GroupMappingIterator() = default;
+        virtual uint16_t Count()        = 0;
+        virtual bool HasNext()          = 0;
+        virtual GroupId Next() const    = 0;
+
+    protected:
+        GroupMappingIterator() = default;
+    };
+
+    class GroupStateIterator
+    {
+    public:
+        virtual ~GroupStateIterator()           = default;
+        virtual uint16_t Count()                = 0;
+        virtual bool HasNext()                  = 0;
+        virtual const GroupState * Next() const = 0;
+
+    protected:
+        GroupStateIterator() = default;
+    };
+
+    class KeySetIterator
+    {
+    public:
+        virtual ~KeySetIterator()           = default;
+        virtual uint16_t Count()            = 0;
+        virtual bool HasNext()              = 0;
+        virtual const KeySet * Next() const = 0;
+
+    protected:
+        KeySetIterator() = default;
+    };
+
+    struct GroupListener
+    {
+        void OnStateChanged(GroupState & old_state, GroupState & new_state);
```

Comment:
This should be converted to either a pure interface or function signatures with also a void* context arg.

GroupListener should have `OnGroupRemoved` as well.